src/objective-c/GRPCClient/GRPCInterceptor.h

```diff
@@ -0,0 +1,269 @@
+ *
+ *
+ *                                   -----------
+ *                                  | GRPCCall2 |
+ *                                   -----------
+ *                                        |
+ *                                        |
+ *                           --------------------------
+ *                          | GRPCInterceptorManager 1 |
+ *                           --------------------------
+ *                          | GRPCInterceptor 1        |
+ *                           --------------------------
+ *                                        |
+ *                                       ...
+ *                                        |
+ *                           --------------------------
+ *                          | GRPCInterceptorManager N |
+ *                           --------------------------
+ *                          | GRPCInterceptor N        |
+ *                           --------------------------
+ *                                        |
+ *                                        |
+ *                               ------------------
+ *                              | GRPCCallInternal |
+ *                               ------------------
+ *
+ * The chain of interceptors is initialized when the corresponding GRPCCall2 object or proto call
+ * object (GRPCUnaryProtoCall and GRPCStreamingProtoCall) is initialized. The initialization of the
+ * chain is controlled by the property interceptorFactories in the callOptions parameter of the
+ * corresponding call object. Property interceptorFactories is an array of
+ * id<GRPCInterceptorFactory> objects provided by the user. When a call object is initialized, each
+ * interceptor factory generates an interceptor object for the call. gRPC internally links the
+ * interceptors with each other and with the actual call object. The order of the interceptors in
+ * the chain is exactly the same as the order of factory objects in interceptorFactories property.
+ * All requests (start, write, finish, cancel, receive next) initiated by the user will be processed
+ * in the order of interceptors, and all responses (initial metadata, data, trailing metadata, write
+ * data done) are processed in the reverse order.
+ *
+ * Each interceptor in the interceptor chain should behave as a user of the next interceptor, and at
+ * the same time behave as a call to the previous interceptor. Therefore interceptor implementations
+ * must follow the state transition of gRPC calls and must also forward events that are consistent
+ * with the current state of the next/previous interceptor. They should also make sure that the
+ * events they forwarded to the next and previous interceptors will, in the end, make the neighbour
+ * interceptor terminate correctly and reaches "finished" state. The diagram below shows the state
+ * transitions. Any event not appearing on the diagram means the event is not permitted for that
```

Comment:
Do we have some checks in the code that assert these are the only valid state transitions?
Somewhat related question: why is is not ok to `receiveNextMessages` in half-close state but it's ok to receive that event in finished state?

Reply:
We have check in `GRPCCallInternal`, but we don't have that in an interceptor; the authors of the interceptor need to take that responsibility.

About `receiveNextMessages`, you are right that we should allow `receiveNextmessages` in half-close state.